### PR TITLE
fix(remix-dev/cli): Remove excess fetching and improve error handling

### PR DESCRIPTION
**Pull request by chaance**
Instead of fetching early to detect template types in the CLI, we can simplify the code and speed up the setup process with a few heuristics:

1. Check if the user passed a local file. If they hand us an explicit file URL, we'll validate it first. Otherwise we just ping the filesystem to see if the string references a filepath and, if not, move on.
2. If the provided template contains no slashes, spaces, or special chars, we assume it is one of our templates and attempt to fetch it.
3. If the template is a GitHub URL or shorthand :org/:repo, we fetch the repo.
4. Any other URLs are assumed to be direct tarball URLs, so we try to fetch them.

Worth noting that there are probably more opportunities to improve error handling. We'll want to iterate here.

- [ ] Docs
- [x] Tests

### CODE REVIEW

File: packages/remix-dev/cli/create.ts [hunk 1]
```diff
@@ -51,9 +64,10 @@ export async function createApp({
   let projectDirIsCurrentDir = relativeProjectDir === "";
   if (!projectDirIsCurrentDir) {
     if (fse.existsSync(projectDir)) {
-      throw new Error(
+      console.error(
```
Comment (chaance) [inline packages/remix-dev/cli/create.ts:67]: Throwing known errors barfs out a call stack that isn't really useful to users. We know the problem, so let's just tell them and exit. They will open an issue if there's a problem and we can deal with the call stack. May want to pass the `--debug` flag here if they really want it.

File: packages/remix-dev/cli/create.ts [hunk 2]
```diff
@@ -433,24 +544,66 @@ type TemplateType =
...
-  template: string,
-  useTypeScript: boolean,
-  token?: string
-): Promise<TemplateType> {
-  if (template.startsWith("file://") || fse.existsSync(template)) {
-    return "local";
+function detectTemplateType(template: string): TemplateType {
+  // 1. Check if the user passed a local file. If they hand us an explicit file
+  //    URL, we'll validate it first. Otherwise we just ping the filesystem to
+  //    see if the string references a filepath and, if not, move on.
```
Comment (brophdawg11) [inline packages/remix-dev/cli/create.ts:550]: This ordered flow makes sense to me - nice work 👍

File: packages/remix-dev/cli/create.ts (hunk 1 above)
Comment (mcansh) [reply packages/remix-dev/cli/create.ts:67]: this technically doesn't have to be used only by the CLI which was the reason for throwing and having the CLI catch them and convert it to a log and process.exit

Comment (chaance) [reply packages/remix-dev/cli/create.ts:67]: Hmm, do we document this somewhere? I don't know that an export should constitute public API but we can chat about that. Happy to change it if we're catching them elsewhere, but currently if a user hits any error they get all the logs.

File: packages/remix-dev/cli/create.ts (hunk 2 above)
Comment (kentcdodds) [reply packages/remix-dev/cli/create.ts:589]: Agreed 💯

**Review by mcansh — APPROVED** [review]

File: packages/remix-dev/cli/run.ts [hunk 3]
```diff
@@ -278,6 +296,8 @@ export async function run(argv: string[] = process.argv.slice(2)) {
 
       await commands.create({
         appTemplate: flags.template ?? answers.appTemplate,
+        templateType:
+          templateType! || (answers.appType === "stack" ? "repo" : "template"),
         projectDir,
         remixVersion: flags.remixVersion,
         installDeps: flags.install ?? answers.install,
```
Comment (mcansh) [inline packages/remix-dev/cli/run.ts:303]: can you change these to logical ORs and I'll close #2440?
using nullish coalescing prevents our CLI from showing the minimum node version error
```suggestion
        installDeps: flags.install || answers.install,
        useTypeScript: flags.typescript || answers.useTypeScript,
```

File: packages/remix-dev/cli/run.ts [hunk 4]
```diff
@@ -305,7 +305,8 @@ export async function run(argv: string[] = process.argv.slice(2)) {
           templateType! || (answers.appType === "stack" ? "repo" : "template"),
         projectDir,
         remixVersion: flags.remixVersion,
-        installDeps: flags.install ?? answers.install,
+        installDeps: flags.install || answers.install,
+        useTypeScript: flags.typescript || answers.useTypeScript,
         useTypeScript: flags.typescript ?? answers.useTypeScript,
```
Comment (mcansh) [inline packages/remix-dev/cli/run.ts:310]: heh oops, the commit suggestion didn't let me remove the previous `useTypeScript` 🤦

File: packages/remix-dev/cli/run.ts [hunk 5]
```diff
@@ -1,11 +1,12 @@
 import * as path from "path";
-import meow from "meow";
 import inspector from "inspector";
+import meow from "meow";
 import inquirer from "inquirer";
-// import chalkAnimation from "chalk-animation";
```
Comment (kentcdodds) [inline packages/remix-dev/cli/run.ts:5]: Could you remove this dep from the lockfile and package.json so we can close #2547?

**Comment by kiliman** [comment]
Hmm.. just noticed that `create-remix@latest` doesn't even announce that you're creating a Remix project. Also, a previous version stopped displaying the version as well.

Any chance we can get those back in the CLI? It's nice knowing what version I'm going to be installing.

**Comment by kentcdodds** [comment]
Once we get this merged, I'm going to be doing a bit more work in the CLI too. I may bring back the animated thing as well.

**Comment by mcansh** [comment]
> Once we get this merged, I'm going to be doing a bit more work in the CLI too. I may bring back the animated thing as well.

sweet! i loved that little touch

**Comment by brophdawg11** [comment]
> I may bring back the animated thing as well.

yes please!! 🙌

**Comment by chaance** [comment]
Note about the animation: I like it as well but it creates some noise for screen-reader users. Want to make sure we ensure that the --no-color flag is respected there.

**Pull request merged**